man/mosquitto.8
'\" -*- coding: us-ascii -*-
.if \n(.g .ds T< \\FC
.if \n(.g .ds T> \\F[\n[.fam]]
.de URL
\\$2 \(la\\$1\(ra\\$3
..
.if \n(.g .mso www.tmac
.TH mosquitto 8 "3 December 2010" "" ""
.SH NAME
mosquitto \- an mqtt broker
.SH SYNOPSIS
'nh
.fi
.ad l
\fBmosquitto\fR \kx
.if (\nx>(\n(.l/2)) .nr x (\n(.l/5)
'in \n(.iu+\nxu
[\fB-c\fR \fIconfig file\fR] [\fB-d\fR | \fB--daemon\fR] [\fB-p\fR \fIport number\fR]
'in \n(.iu-\nxu
.ad b
'hy
.SH DESCRIPTION
\fBmosquitto\fR is a broker for the mqtt protocol version 3.
.PP
It is intended to be a drop in replacement for rsmb, but it's a little way off yet.
.SH OPTIONS
.TP 
\*(T<\fB\-c\fR\*(T>, \*(T<\fB\-\-config\-file\fR\*(T>
Load configuration from a file. If not given, the default values as described in \fBmosquitto.conf\fR(5) are used.
.TP 
\*(T<\fB\-d\fR\*(T>, \*(T<\fB\-\-daemon\fR\*(T>
Run \fBmosquitto\fR in the background as a daemon. All other behaviour remains the same.
.TP 
\*(T<\fB\-p\fR\*(T>, \*(T<\fB\-\-port\fR\*(T>
Listen on the port specified instead of the default 1883. This acts in addition to the port setting in the config file. May be specified multiple times to open multiple sockets listening on different ports. This socket will be bound to all network interfaces.
.SH LIMITATIONS
\fBmosquitto\fR implements all of the mqtt protocol version 3, but there are some significant limitations compared to rsmb.
.TP 0.2i
\(bu
There is only basic bridge support.
.TP 0.2i
\(bu
There are fewer configuration options.
.TP 0.2i
\(bu
There is less logging information.
.TP 0.2i
\(bu
The $SYS hierarchy is smaller.
.TP 0.2i
\(bu
It isn't as well tested or documented...
.PP
It should be noted that all of the above limitations should have "currently" included in them. They will all be fixed at some point.
.SH "BROKER STATUS"
Clients can find information about the broker by subscribing to topics in the $SYS hierarchy as follows. Topics marked as static are only sent once per client on subscription. All other topics are updated every \*(T<\fBsys_interval\fR\*(T> seconds. If \*(T<\fBsys_interval\fR\*(T> is 0, then updates are not sent.
.TP 
\*(T<\fB$SYS/broker/bytes/received\fR\*(T>
The total number of bytes received since the broker started.
.TP 
\*(T<\fB$SYS/broker/bytes/sent\fR\*(T>
The total number of bytes sent since the broker started.
.TP 
\*(T<\fB$SYS/broker/changeset\fR\*(T>
The repository changeset (revision) associated with this build.
.TP 
\*(T<\fB$SYS/broker/clients/total\fR\*(T>
The total number of clients currently either connected to the broker or disconnected but with clean session not set.
.TP 
\*(T<\fB$SYS/broker/heap/current size\fR\*(T>
The current size of the heap memory in use by
mosquitto. Note that this topic may be unavailable
depending on compile time options.
.TP 
\*(T<\fB$SYS/broker/messages/stored\fR\*(T>
The number of messages currently held in the message store.
.TP 
\*(T<\fB$SYS/broker/messages/received\fR\*(T>
The total number of messages received since the broker started.
.TP 
\*(T<\fB$SYS/broker/messages/sent\fR\*(T>
The total number of messages sent since the broker started.
.TP 
\*(T<\fB$SYS/broker/timestamp\fR\*(T>
The timestamp at which this particular build of the broker was made. Static.
.TP 
\*(T<\fB$SYS/broker/uptime\fR\*(T>
The amount of time in seconds the broker has been online.
.TP 
\*(T<\fB$SYS/broker/version\fR\*(T>
The version of the broker. Static.
.SH "WILDCARD TOPIC SUBSCRIPTIONS"
In addition to allowing clients to subscribe to specific topics,
mosquitto also allows the use of two wildcards in subscriptions.
\*(T<\fB+\fR\*(T> is the wildcard used to match a single level of
hierarchy. For example, for a topic of "a/b/c/d", the following example
subscriptions will match:
.TP 0.2i
\(bu
a/b/c/d
.TP 0.2i
\(bu
+/b/c/d
.TP 0.2i
\(bu
a/+/c/d
.TP 0.2i
\(bu
a/+/+/d
.TP 0.2i
\(bu
+/+/+/+
.PP
The following subscriptions will not match:
.TP 0.2i
\(bu
a/b/c
.TP 0.2i
\(bu
b/+/c/d
.TP 0.2i
\(bu
+/+/+
.PP
The second wildcard is \*(T<\fB#\fR\*(T> and is used to match
all subsequent levels of hierarchy. With a topic of "a/b/c/d", the
following example subscriptions will match:
.TP 0.2i
\(bu
a/b/c/d
.TP 0.2i
\(bu
#
.TP 0.2i
\(bu
a/#
.TP 0.2i
\(bu
a/b/#
.TP 0.2i
\(bu
a/b/c/#
.TP 0.2i
\(bu
+/b/c/#
.PP
The $SYS hierarchy does not match a subscription of
"#". If you want to observe the entire $SYS hierarchy, subscribe to
$SYS/#.
.PP
Note that the wildcards must be only ever used on their own, so a
subscription of "a/b+/c" is not valid use of a wildcard. The
\*(T<\fB#\fR\*(T> wildcard must only ever be used as the final
character of a subscription.
.SH BRIDGES
Multiple brokers can be connected together with the bridging
functionality. This is useful where it is desirable to share
information between locations, but where not all of the information
needs to be shared. An example could be where a number of users are
running a broker to help record power usage and for a number of other
reasons. The power usage could be shared through bridging all of the
user brokers to a common broker, allowing the power usage of all
users to be collected and compared. The other information would remain
local to each
broker.
.PP
For information on configuring bridges, see
\fBmosquitto.conf\fR(5).
.SH SIGNALS
.TP 
SIGUSR1
Upon receiving the SIGUSR1 signal, mosquitto will write the persistence database to disk. This signal is only acted upon if persistence is enabled.
.TP 
SIGUSR2
The SIGUSR2 signal causes mosquitto to run VACUUM on
the in-memory database. This frees currently unused memory
associated with the database.
.SH FILES
.TP 
\*(T<\fI/etc/mosquitto.conf\fR\*(T>
Configuration file. See \fBmosquitto.conf\fR(5).
.TP 
\*(T<\fI/var/lib/mosquitto/mosquitto.db\fR\*(T>
Persistent message data storage location if persist enabled.
.TP 
\*(T<\fI/etc/hosts.allow\fR\*(T>, \*(T<\fI/etc/hosts.deny\fR\*(T>
Host access control via tcp-wrappers as described in \fBhosts_access\fR(5).
.SH BUGS
\fBmosquitto\fR bug information can be found at http://launchpad.net/mosquitto.
.SH "SEE ALSO"
\fBmosquitto.conf\fR(5)
\fBhosts_access\fR(5)
\fBmosquitto_pub\fR(1)
\fBmosquitto_sub\fR(1)
\fBlibmosquitto\fR(3)
.SH THANKS
Thanks to Andy Stanford-Clark for being one of the people who came up with MQTT in the first place and providing clarifications of the protocol.
.PP
Thanks also to everybody at the Ubuntu UK Podcast and Linux Outlaws for organising OggCamp, where Andy gave a talk that inspired mosquitto.
.SH AUTHOR
Roger Light <\*(T<roger@atchoo.org\*(T>>


src/mosquitto.c
/*
Copyright (c) 2009,2010, Roger Light <roger@atchoo.org>
All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

1. Redistributions of source code must retain the above copyright notice,
   this list of conditions and the following disclaimer.
2. Redistributions in binary form must reproduce the above copyright
   notice, this list of conditions and the following disclaimer in the
   documentation and/or other materials provided with the distribution.
3. Neither the name of mosquitto nor the names of its
   contributors may be used to endorse or promote products derived from
   this software without specific prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
POSSIBILITY OF SUCH DAMAGE.
*/

#include <config.h>

#ifndef WIN32
#include <poll.h>
#include <pwd.h>
#include <unistd.h>
#else
#include <process.h>
#include <winsock2.h>
#include <ws2tcpip.h>
#endif

#include <errno.h>
#include <signal.h>
#include <stdio.h>
#include <string.h>
#include <sys/stat.h>
#ifdef WITH_WRAP
#include <tcpd.h>
#endif

#include <mqtt3.h>
#include <memory_mosq.h>

static int run;
#ifdef WITH_WRAP
#include <syslog.h>
int allow_severity = LOG_INFO;
int deny_severity = LOG_INFO;
#endif

mosquitto_db int_db;

int drop_privileges(mqtt3_config *config);
void handle_sigint(int signal);
void handle_sigusr1(int signal);
void handle_sigusr2(int signal);
static void loop_handle_errors(struct pollfd *pollfds);
static void loop_handle_reads_writes(struct pollfd *pollfds);

/* mosquitto shouldn't run as root.
 * This function will attempt to change to an unprivileged user and group if
 * running as root. The user is given in config->user.
 * Returns 1 on failure (unknown user, setuid/setgid failure)
 * Returns 0 on success.
 * Note that setting config->user to "root" does not produce an error, but it
 * strongly discouraged.
 */
int drop_privileges(mqtt3_config *config)
{
#if !defined(__CYGWIN__) && !defined(WIN32)
	struct passwd *pwd;

	if(geteuid() == 0){
		if(config->user){
			pwd = getpwnam(config->user);
			if(!pwd){
				mqtt3_log_printf(MOSQ_LOG_ERR, "Error: Invalid user '%s'.", config->user);
				return 1;
			}
			if(setgid(pwd->pw_gid) == -1){
				mqtt3_log_printf(MOSQ_LOG_ERR, "Error: %s.", strerror(errno));
				return 1;
			}
			if(setuid(pwd->pw_uid) == -1){
				mqtt3_log_printf(MOSQ_LOG_ERR, "Error: %s.", strerror(errno));
				return 1;
			}
		}
		if(geteuid() == 0 || getegid() == 0){
			mqtt3_log_printf(MOSQ_LOG_WARNING, "Warning: Mosquitto should not be run as root/administrator.");
		}
	}
#endif
	return MOSQ_ERR_SUCCESS;
}

int loop(mqtt3_config *config, int *listensock, int listensock_count, int listener_max)
{
	time_t start_time = time(NULL);
	time_t last_backup = time(NULL);
	time_t last_store_clean = time(NULL);
	time_t now;
	int fdcount;
#ifndef WIN32
	sigset_t sigblock, origsig;
#endif
	int i;
	struct pollfd *pollfds = NULL;
	unsigned int pollfd_count = 0;
	int new_clients = 1;
	int client_max = 0;
	unsigned int sock_max = 0;

#ifndef WIN32
	sigemptyset(&sigblock);
	sigaddset(&sigblock, SIGINT);
#endif

	while(run){
		mqtt3_db_sys_update(&int_db, config->sys_interval, start_time);

		if(new_clients){
			client_max = -1;
			for(i=0; i<int_db.context_count; i++){
				if(int_db.contexts[i] && int_db.contexts[i]->core.sock != -1 && int_db.contexts[i]->core.sock > sock_max){
					client_max = int_db.contexts[i]->core.sock;
				}
			}
			new_clients = 0;
		}

		if(client_max > listener_max){
			sock_max = client_max;
		}else{
			sock_max = listener_max;
		}
		if(sock_max+1 > pollfd_count){
			pollfd_count = sock_max+1;
			pollfds = _mosquitto_realloc(pollfds, sizeof(struct pollfd)*pollfd_count);
			if(!pollfds){
				mqtt3_log_printf(MOSQ_LOG_ERR, "Error: Out of memory.");
				return MOSQ_ERR_NOMEM;
			}
		}

		memset(pollfds, -1, sizeof(struct pollfd)*pollfd_count);

		for(i=0; i<listensock_count; i++){
			pollfds[listensock[i]].fd = listensock[i];
			pollfds[listensock[i]].events = POLLIN;
			pollfds[listensock[i]].revents = 0;
		}

		now = time(NULL);
		for(i=0; i<int_db.context_count; i++){
			if(int_db.contexts[i]){
				if(int_db.contexts[i]->core.sock != -1){
					if(int_db.contexts[i]->bridge){
						mqtt3_check_keepalive(int_db.contexts[i]);
					}
					if(!(int_db.contexts[i]->core.keepalive) || now - int_db.contexts[i]->core.last_msg_in < (time_t)(int_db.contexts[i]->core.keepalive)*3/2){
						if(mqtt3_db_message_write(int_db.contexts[i])){
							// FIXME - do something here.
						}
						pollfds[int_db.contexts[i]->core.sock].fd = int_db.contexts[i]->core.sock;
						pollfds[int_db.contexts[i]->core.sock].events = POLLIN;
						pollfds[int_db.contexts[i]->core.sock].revents = 0;
						if(int_db.contexts[i]->core.out_packet){
							pollfds[int_db.contexts[i]->core.sock].events |= POLLOUT;
						}
					}else{
						mqtt3_log_printf(MOSQ_LOG_NOTICE, "Client %s has exceeded timeout, disconnecting.", int_db.contexts[i]->core.id);
						/* Client has exceeded keepalive*1.5 */
						if(int_db.contexts[i]->bridge || int_db.contexts[i]->clean_session == false){
							mqtt3_socket_close(int_db.contexts[i]);
						}else{
							mqtt3_context_cleanup(&int_db, int_db.contexts[i], true);
							int_db.contexts[i] = NULL;
						}
					}
				}else{
					if(int_db.contexts[i]->bridge){
						/* Want to try to restart the bridge connection */
						if(!int_db.contexts[i]->bridge->restart_t){
							int_db.contexts[i]->bridge->restart_t = time(NULL)+30;
						}else{
							if(time(NULL) > int_db.contexts[i]->bridge->restart_t){
								int_db.contexts[i]->bridge->restart_t = 0;
								mqtt3_bridge_connect(&int_db, int_db.contexts[i]);
							}
						}
					}else{
						if(int_db.contexts[i]->clean_session == true){
							mqtt3_context_cleanup(&int_db, int_db.contexts[i], true);
							int_db.contexts[i] = NULL;
						}
					}
				}
			}
		}

		mqtt3_db_message_timeout_check(&int_db, config->retry_interval);

#ifndef WIN32
		sigprocmask(SIG_SETMASK, &sigblock, &origsig);
		fdcount = poll(pollfds, pollfd_count, 1000);
		sigprocmask(SIG_SETMASK, &origsig, NULL);
#else
		fdcount = WSAPoll(pollfds, pollfd_count, 1000);
#endif
		if(fdcount == -1){
			loop_handle_errors(pollfds);
		}else{
			loop_handle_reads_writes(pollfds);

			for(i=0; i<listensock_count; i++){
				if(pollfds[listensock[i]].revents & (POLLIN | POLLPRI)){
					new_clients = 1;
					while(mqtt3_socket_accept(&int_db.contexts, &int_db.context_count, listensock[i]) != -1){
					}
				}
			}
		}
#ifdef WITH_PERSISTENCE
		if(config->persistence && config->autosave_interval){
			if(last_backup + config->autosave_interval < now){
				mqtt3_db_backup(&int_db, false, false);
				last_backup = time(NULL);
			}
		}
#endif
		if(!config->store_clean_interval || last_store_clean + config->store_clean_interval < now){
			mqtt3_db_store_clean(&int_db);
			last_store_clean = time(NULL);
		}
	}

	if(pollfds) _mosquitto_free(pollfds);
	return MOSQ_ERR_SUCCESS;
}

/* Error ocurred, probably an fd has been closed. 
 * Loop through and check them all.
 */
static void loop_handle_errors(struct pollfd *pollfds)
{
	int i;

	for(i=0; i<int_db.context_count; i++){
		if(int_db.contexts[i] && int_db.contexts[i]->core.sock != -1){
			if(pollfds[int_db.contexts[i]->core.sock].revents & POLLHUP){
				if(int_db.contexts[i]->core.state != mosq_cs_disconnecting){
					mqtt3_log_printf(MOSQ_LOG_NOTICE, "Socket error on client %s, disconnecting.", int_db.contexts[i]->core.id);
					mqtt3_db_client_will_queue(&int_db, int_db.contexts[i]);
				}else{
					mqtt3_log_printf(MOSQ_LOG_NOTICE, "Client %s disconnected.", int_db.contexts[i]->core.id);
				}
				if(int_db.contexts[i]->bridge || int_db.contexts[i]->clean_session == false){
					mqtt3_socket_close(int_db.contexts[i]);
				}else{
					mqtt3_context_cleanup(&int_db, int_db.contexts[i], true);
					int_db.contexts[i] = NULL;
				}
			}
		}
	}
}

static void loop_handle_reads_writes(struct pollfd *pollfds)
{
	int i;

	for(i=0; i<int_db.context_count; i++){
		if(int_db.contexts[i] && int_db.contexts[i]->core.sock != -1){
			if(pollfds[int_db.contexts[i]->core.sock].revents & POLLOUT){
				if(mqtt3_net_write(int_db.contexts[i])){
					if(int_db.contexts[i]->core.state != mosq_cs_disconnecting){
						mqtt3_log_printf(MOSQ_LOG_NOTICE, "Socket write error on client %s, disconnecting.", int_db.contexts[i]->core.id);
						mqtt3_db_client_will_queue(&int_db, int_db.contexts[i]);
					}else{
						mqtt3_log_printf(MOSQ_LOG_NOTICE, "Client %s disconnected.", int_db.contexts[i]->core.id);
					}
					/* Write error or other that means we should disconnect */
					/* Bridges don't get cleaned up because they will reconnect later. */
					if(int_db.contexts[i]->bridge || int_db.contexts[i]->clean_session == false){
						mqtt3_socket_close(int_db.contexts[i]);
					}else{
						mqtt3_context_cleanup(&int_db, int_db.contexts[i], true);
						int_db.contexts[i] = NULL;
					}
				}
			}
		}
		if(int_db.contexts[i] && int_db.contexts[i]->core.sock != -1){
			if(pollfds[int_db.contexts[i]->core.sock].revents & POLLIN){
				if(mqtt3_net_read(&int_db, int_db.contexts[i])){
					if(int_db.contexts[i]->core.state != mosq_cs_disconnecting){
						mqtt3_log_printf(MOSQ_LOG_NOTICE, "Socket read error on client %s, disconnecting.", int_db.contexts[i]->core.id);
						mqtt3_db_client_will_queue(&int_db, int_db.contexts[i]);
					}else{
						mqtt3_log_printf(MOSQ_LOG_NOTICE, "Client %s disconnected.", int_db.contexts[i]->core.id);
					}
					/* Read error or other that means we should disconnect */
					/* Bridges don't get cleaned up because they will reconnect later. */
					if(int_db.contexts[i]->bridge || int_db.contexts[i]->clean_session == false){
						mqtt3_socket_close(int_db.contexts[i]);
					}else{
						mqtt3_context_cleanup(&int_db, int_db.contexts[i], true);
						int_db.contexts[i] = NULL;
					}
				}
			}
		}
	}
}

/* Signal handler for SIGINT and SIGTERM - just stop gracefully. */
void handle_sigint(int signal)
{
	run = 0;
}

/* Signal handler for SIGUSR1 - backup the db. */
void handle_sigusr1(int signal)
{
#ifdef WITH_PERSISTENCE
	mqtt3_db_backup(&int_db, false, false);
#endif
}

/* Signal handler for SIGUSR2 - vacuum the db. */
void handle_sigusr2(int signal)
{
	//mqtt3_db_vacuum();
	mqtt3_sub_tree_print(&int_db.subs, 0);
}

int main(int argc, char *argv[])
{
	int *listensock = NULL;
	int listensock_count = 0;
	int listensock_index = 0;
	int *socks, sock_count;
	mqtt3_config config;
	char buf[1024];
	int i, j;
	FILE *pid;
	int listener_max;
	int rc;

#ifdef WIN32
	WSADATA wsaData;
	WSAStartup(MAKEWORD(2,2), &wsaData);
#endif

	mqtt3_config_init(&config);
	if(mqtt3_config_parse_args(&config, argc, argv)) return 1;

	if(config.daemon){
#ifndef WIN32
		switch(fork()){
			case 0:
				break;
			case -1:
				mqtt3_log_printf(MOSQ_LOG_ERR, "Error in fork: %s", strerror(errno));
				return 1;
			default:
				return MOSQ_ERR_SUCCESS;
		}
#else
		mqtt3_log_printf(MOSQ_LOG_WARNING, "Warning: Can't start in daemon mode in Windows.");
#endif
	}

	if(config.daemon && config.pid_file){
		pid = fopen(config.pid_file, "wt");
		if(pid){
			fprintf(pid, "%d", getpid());
			fclose(pid);
		}else{
			mqtt3_log_printf(MOSQ_LOG_ERR, "Error: Unable to write pid file.");
			return 1;
		}
	}
	if(drop_privileges(&config)) return 1;

	if(mqtt3_db_open(&config, &int_db)){
		mqtt3_log_printf(MOSQ_LOG_ERR, "Error: Couldn't open database.");
		return 1;
	}
	/* Initialise logging only after initialising the database in case we're
	 * logging to topics */
	mqtt3_log_init(config.log_type, config.log_dest);
	mqtt3_log_printf(MOSQ_LOG_INFO, "mosquitto version %s (build date %s) starting", VERSION, TIMESTAMP);

	/* Set static $SYS messages */
	snprintf(buf, 1024, "mosquitto version %s", VERSION);
	mqtt3_db_messages_easy_queue(&int_db, NULL, "$SYS/broker/version", 2, strlen(buf), (uint8_t *)buf, 1);
	snprintf(buf, 1024, "%s", TIMESTAMP);
	mqtt3_db_messages_easy_queue(&int_db, NULL, "$SYS/broker/timestamp", 2, strlen(buf), (uint8_t *)buf, 1);
	snprintf(buf, 1024, "%s", "$Revision: 9e9ae08f60d5 $"); // Requires hg keyword extension.
	mqtt3_db_messages_easy_queue(&int_db, NULL, "$SYS/broker/changeset", 2, strlen(buf), (uint8_t *)buf, 1);
	snprintf(buf, 1024, "%d-bit", 8*sizeof(dbid_t));
	mqtt3_db_messages_easy_queue(&int_db, NULL, "$SYS/broker/dbid_size", 2, strlen(buf), (uint8_t *)buf, 1);

	listener_max = -1;
	listensock_index = 0;
	for(i=0; i<config.listener_count; i++){
		if(mqtt3_socket_listen(config.listeners[i].host, config.listeners[i].port, &socks, &sock_count)){
			_mosquitto_free(int_db.contexts);
			mqtt3_db_close(&int_db);
			if(config.pid_file){
				remove(config.pid_file);
			}
			return 1;
		}
		listensock_count += sock_count;
		listensock = _mosquitto_realloc(listensock, sizeof(int)*listensock_count);
		if(!listensock){
			_mosquitto_free(int_db.contexts);
			mqtt3_db_close(&int_db);
			if(config.pid_file){
				remove(config.pid_file);
			}
			return 1;
		}
		for(j=0; j<sock_count; j++){
			if(socks[j] < 0){
				_mosquitto_free(int_db.contexts);
				mqtt3_db_close(&int_db);
				if(config.pid_file){
					remove(config.pid_file);
				}
				return 1;
			}
			listensock[listensock_index] = socks[j];
			if(listensock[listensock_index] > listener_max){
				listener_max = listensock[listensock_index];
			}
			listensock_index++;
		}
		_mosquitto_free(socks);
	}

	signal(SIGINT, handle_sigint);
	signal(SIGTERM, handle_sigint);
#ifndef WIN32
	signal(SIGUSR1, handle_sigusr1);
	signal(SIGUSR2, handle_sigusr2);
	signal(SIGPIPE, SIG_IGN);
#endif

	for(i=0; i<config.bridge_count; i++){
		if(mqtt3_bridge_new(&int_db, &(config.bridges[i]))){
			mqtt3_log_printf(MOSQ_LOG_WARNING, "Warning: Unable to connect to bridge %s.", 
					config.bridges[i].name);
		}
	}
	run = 1;
	rc = loop(&config, listensock, listensock_count, listener_max);

	mqtt3_log_printf(MOSQ_LOG_INFO, "mosquitto version %s terminating", VERSION);
	mqtt3_log_close();

#ifdef WITH_PERSISTENCE
	if(config.persistence && config.autosave_interval){
		mqtt3_db_backup(&int_db, true, true);
	}
#endif

	for(i=0; i<int_db.context_count; i++){
		if(int_db.contexts[i]){
			mqtt3_context_cleanup(&int_db, int_db.contexts[i], true);
		}
	}
	_mosquitto_free(int_db.contexts);
	int_db.contexts = NULL;
	mqtt3_db_close(&int_db);

	if(listensock){
		for(i=0; i<listensock_count; i++){
			if(listensock[i] != -1){
#ifndef WIN32
				close(listensock[i]);
#else
				closesocket(listensock[i]);
#endif
			}
		}
		_mosquitto_free(listensock);
	}


	if(config.pid_file){
		remove(config.pid_file);
	}

#ifdef WIN32
	WSACleanup();
#endif

	return rc;
}

#ifdef WIN32
int __stdcall WinMain(HINSTANCE hInstance, HINSTANCE hPrevInstance, LPSTR lpCmdLine, int nCmdShow)
{
	char *argv = "mosquitto";
	return main(1, &argv);
}
#endif
